### DISCUSSION TITLE
How to connect with prisma to a azure postgres flexible server in a secure way?

### DISCUSSION THREAD

**Discussion by andreasvh-conceto** [prisma/prisma#13462] [Q&A] [answered]
Hello,
currently we connect to a Microsoft Azure Postgres Flexible server via the internet traffic. In production it has to be encrypted and man in the middle attacks should be avoided. Microsoft Azure recommends the [verify-full](https://docs.microsoft.com/en-us/azure/postgresql/single-server/concepts-ssl-connection-security#connect-using-psql) sslmode for that. In prisma docs i can only see the required parameter, which i use and set up to require.

`DATABASE_URL="postgresql://<username>:<password>@<servername>.postgres.database.azure.com/<databaseName>?schema=<schemaName>&sslmode=require"`

Prisma Client is working out of the box with this url, but how can i be sure, that the certificate from the server is verified? Can i be sure, that the traffic is encrypted and the chain of trust is verified? I wonder, because i did not set up any ca certificates (like the [prisma docs](https://www.prisma.io/docs/concepts/components/prisma-schema/data-sources#securing-database-connections) says in the url and would expect an error. 

Microsoft docs say to download the their root certificate for postgres here: [Azure Postgres Root CA](https://docs.microsoft.com/en-us/azure/postgresql/single-server/concepts-ssl-connection-security#applications-that-require-certificate-verification-for-tls-connectivity)
I can also download the root certificate in the network area of our postgres sql flexible server. Any recommendation, who to set this up in the right way?

Thanks a lot for your help and best regards,
Andreas

**Comment by pantharshit00** [comment]
Hello @andreasvh-conceto, our apologies for the delay in getting back to you! Due to the large number of questions in our community channels, unfortunately this slipped through the cracks.

If you want Client verification of the server certificates, then you will need to follow the Prisma docs and setup the certificates according to the instruction there. It will work in sslmode without providing certificates to the client but the client won't compare those against a stored copy.

**Reply by vladi-stevanovic** [reply]
:wave: Hello @andreasvh-conceto, I’m going through pending GitHub Discussions to check if you need any help. Is this issue still open?

I see another user recommended a [potential workaround](https://github.com/prisma/prisma/discussions/13462#discussioncomment-4144745). Please let us know you still need help! :raised_hands:

**Comment by Dr1Ku** [comment]
Not _very_ related but I can share what solved my woes when trying to use `prisma` with a no-frills, regular 
`Azure Database for PostgreSQL single server`. The connection would apparently work but I still got 
`P1010: User <DB-user> was denied access on the database <DB>.public`. I thought it was a permissions
/`grant`s  issue, then that it was a firewall issue ([see also](https://stackoverflow.com/a/68535786)), then that maybe I do need a shadow database 
connection as well (not needed since I'm just trying to `push`). What also helped was to review the errors
that are being thrown around via a DB-specific tool (in this case, say, `pgAdmin` but `DBeaver`/`DataGrip` 
are also of good help). `prisma`, being generic, wouldn't display _that_ specific error messages, which made
it a bit hard to debug.

Anyhow, here's what worked in my case:

0. Log in with your Azure-provided DB Admin credentials (via e.g. a PostgresSQL CLI/tool of your choice)
1. Create your DB and Role/User
2. Be sure to heed the following DB connection string gotchas:
  * Yes, you have to fully specify `user@short-host`, whereby 'short-host' (thought-out name) is the part of the 
    regular/long Azure DB URI (`<short-host>.postgres.database.azure.com`).
  * Yes, you have to specify a value of `require` for `sslmode`
  * Yes, the newly created Role/User _should_ own the newly created DB

Here's a connection string that worked for an e.g. `npm x prisma -- db push`:  
`postgresql://<user>@<short-host>:<pass>@<short-host>.postgres.database.azure.com:5432/<db>?sslmode=require`

**Comment by andreasvh-conceto** [comment, marked as answer]
We found a solution for that issue - also with using pg bouncer this worked fine. In the security section in azure portal we downloaded the certificate DigiCertGlobalRootCA.azure.prod.crt.pem and put the full path to the sslcert property. Setting sslaccept to strict ensures, that you only trust the certificates signed by this CA. Pg bouncer is just for productive use case and had nothing to do with the initial issue.

```
# Database URL
DATABASE_URL="postgres://<user>:${userPassword}@<dbHost>:6432/<databaseName>?schema=public&connect_timeout=20&sslmode=require&sslcert=/path/to/cert/DigiCertGlobalRootCA.azure.prod.crt.pem&sslaccept=strict&&pgbouncer=true"
```

**Reply by Anishpras** [reply]
@andreasvh-conceto  Do we need to pass pgbouncer=true if we are already mentioning the port of 6432 ?

And using the above URL structure I am getting this error -

```
`prisma.user.findUnique()` invocation:
 
 
Error querying the database: FATAL: unsupported startup parameter: search_path

```

**Reply by kumard3** [reply]
> @andreasvh-conceto Do we need to pass pgbouncer=true if we are already mentioning the port of 6432 ?
> 
> And using the above URL structure I am getting this error -
> 
> ```
> `prisma.user.findUnique()` invocation:
>  
>  
> Error querying the database: FATAL: unsupported startup parameter: search_path
> ```

add this as a parameter 
``ignore_startup_parameters=search_path``

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
